frame/election-provider-multi-phase/src/unsigned.rs

```diff
@@ -144,49 +144,63 @@ impl<T: Config> Pallet<T> {
 			Self::snapshot().ok_or(MinerError::SnapshotUnAvailable)?;
 		let desired_targets = Self::desired_targets().ok_or(MinerError::SnapshotUnAvailable)?;
 
-		// closures.
+		// now make some helper closures.
 		let cache = helpers::generate_voter_cache::<T>(&voters);
 		let voter_index = helpers::voter_index_fn::<T>(&cache);
 		let target_index = helpers::target_index_fn::<T>(&targets);
 		let voter_at = helpers::voter_at_fn::<T>(&voters);
 		let target_at = helpers::target_at_fn::<T>(&targets);
 		let stake_of = helpers::stake_of_fn::<T>(&voters, &cache);
 
-		let ElectionResult { assignments, winners } = election_result;
+		// Compute the size of a compact solution comprised of the selected arguments.
+		//
+		// This function completes in `O(edges)`; it's expensive, but linear.
+		let encoded_size_of = |assignments: &[IndexAssignmentOf<T>]| {
+			CompactOf::<T>::try_from(assignments).map(|compact| compact.encoded_size())
+		};
+
+		let ElectionResult { mut assignments, winners } = election_result;
+		// Sort the assignments by reversed voter stake. This ensures that we can efficiently truncate the list.
+		let stakes: BTreeMap<_, _> =
+			voters.iter().map(|(who, stake, _)| (who.clone(), *stake)).collect();
+		assignments.sort_unstable_by_key(|Assignment::<T> { who, .. }| {
+			sp_std::cmp::Reverse(stakes.get(who).cloned().unwrap_or_default())
+		});
+
+		// Reduce (requires round-trip to staked form)
```

Comment:
yeah I wish we could do this better... something to re-think in long term.

Reply:
Agree, but that was existing code; I didn't want to mess with it for this PR.

Reply:
yeah, all historical garbage :D